File: inventory-service/src/main/java/com/catius/inventory/domain/Inventory.java
package com.catius.inventory.domain;


import com.catius.inventory.exception.InsufficientStockException;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Table
@Getter
@NoArgsConstructor
public class Inventory {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true)
    private Long productId;

    @Column(nullable = false)
    private String productName;

    @Column(nullable = false)
    private int quantity;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    public static Inventory create(Long productId, String productName, int quantity) {
        Inventory inventory = new Inventory();
        inventory.productId = productId;
        inventory.productName = productName;
        inventory.quantity = quantity;
        inventory.createdAt = LocalDateTime.now();
        inventory.updatedAt = LocalDateTime.now();
        return inventory;
    }

    public void deduct(int quantity) {
        if (this.quantity < quantity) {
            throw new InsufficientStockException(productId.toString(), this.quantity, quantity);
        }
        this.quantity -= quantity;
        this.updatedAt = LocalDateTime.now();
    }

    public void restore(int quantity) {
        this.quantity += quantity;
        this.updatedAt = LocalDateTime.now();
    }
}

File: inventory-service/src/main/kotlin/com/catius/inventory/domain/Inventory.kt
package com.catius.inventory.domain

import java.time.Instant

class InsufficientStockException(
    val productId: Long,
    val requested: Int,
    val available: Int,
) : RuntimeException(
    "재고 부족: product=$productId, 요청=$requested, 가용=$available",
)

data class Inventory(
    val productId: Long,
    val availableQuantity: Int,
    val updatedAt: Instant,
) {
    init {
        require(availableQuantity >= 0) {
            "가용 재고는 음수일 수 없습니다: $availableQuantity"
        }
    }

    fun reserve(quantity: Int, at: Instant): Inventory {
        require(quantity > 0) { "차감 수량은 1 이상이어야 합니다: $quantity" }
        if (availableQuantity < quantity) {
            throw InsufficientStockException(productId, quantity, availableQuantity)
        }
        return copy(availableQuantity = availableQuantity - quantity, updatedAt = at)
    }

    fun release(quantity: Int, at: Instant): Inventory {
        require(quantity > 0) { "복원 수량은 1 이상이어야 합니다: $quantity" }
        return copy(availableQuantity = availableQuantity + quantity, updatedAt = at)
    }
}
